### ISSUE TITLE
Is NCCL creating blocking threads?

### ISSUE THREAD

**Issue by frankseide** [NVIDIA/nccl#374]
We are seeing a PyTorch process at CPU load of 400%. We use no multi-threading in the code, so CPU load should be at best 100%. Does NCCL create threads that busy-wait? Our code creates multiple communication groups.

**Comment by sjeaugey** [comment]
Yes, NCCL creates one thread per communicator per rank for network communication. Threads should only be active when there is ongoing NCCL communication, meaning outside of NCCL regions they should be sleeping.

**Comment by frankseide** [comment]
Thanks. Is it possible that they busy-wait, depending on an option? E.g. if CUDA is configured to busy-wait when synchronizing with the GPU, might NCCL obey that setting as well? I am just fishing, we don't have a clear handle where the CPU usage comes from. Unfortunately, we don't have sudo access on the lab machines where we observe it...

**Comment by sjeaugey** [comment]
Yes, it is expected that they busy wait, so seeing 100% CPU for each NCCL thread is expected. Note however that when they have nothing to do, they will call sched_yield() inside the loop, so if anything else needs to use the CPU, their CPU usage should shrink.

**Comment by frankseide** [comment]
We are also seeing n x 100% CPU when there are other processes trying to run (we do a lot of text pre-processing on the fly during model training, which runs in parallel processes). We are seeing slowdowns.

Is there a way to control the scheduling of these threads? E.g. is there a way to place the threads all on the same core, adjust their priority, or switch from sched_yield() (which is a form of self-rolled scheduler) to waiting for a wake-up event?

**Comment by sjeaugey** [comment]
Since those threads use 100% CPU, the kernel might wrongly assume that they should keep a dedicated core. So if your other threads do not use 100% CPU, they might end up sharing a core and NCCL threads might keep a dedicated one.

So you could try to manually pin threads to specific cpus to avoid that (forcing them to share a given CPU). You could also move all NCCL threads to a single core. But in all cases, keep in mind that it could also affect the NCCL performance, especially if NCCL threads share the CPU with other processes which take the CPU for a full 100ms timeslice. If you're using RDMA, having a single core for all NCCL communication might be ok, but if you're using TCP/IP, you might want to use more than one core.

Using wake-up events is not something NCCL can support since communication between the GPU and CPU is based on shared memory, hence uses polling. There are no syscalls in CUDA code. The only alternative would be to use usleep(delay) instead of sched_yield(). The faster your network, the smaller that delay should to avoid hurting performance, so how much CPU you would save would depend on your network speed -- but we haven't tried.

**Comment by frankseide** [comment]
Is there a way to use wake-up events between operations? I.e. have the NCCL thread put itself to sleep between operations, and wake up only while data exchanges are taking place?

**Comment by sjeaugey** [comment]
Between NCCL operations, i.e. when all network transmission is done and the NCCL operation is complete, the thread will go to sleep and be waken up right after launching the next NCCL CUDA kernel.

**Comment by frankseide** [comment]
Thanks for all the information. This will be very valuable in tracking down our slowdown.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
